### PR TITLE
Avoid crash on other Linux due to ref counting issues in Directory cache

### LINKED ISSUE #2148: Crash when typing path in header bar
### What Happened?

When you type some paths in the header bar (e.g. `/home`) and hit enter, it crashes.

```
** (io.elementary.files:4548): CRITICAL **: 22:10:54.221: files_file_set_expanded: assertion 'self->is_directory' failed
[1]    4548 segmentation fault (core dumped)  io.elementary.files
```

### Steps to Reproduce

1. Make sure you are in some subfolder of `/home` (and not `/home` itself), e.g. `/home/bobby285271` or `/home/bobby285271/gala` or something
2. Type `/home` in the header bar
3. Hit <kbd>Enter</kbd>
4. Crash

### Expected Behavior

No crash.

### OS Version

Other Linux

### Software Version

Compiled from git

### Log Output

```shell
#0  g_type_check_instance_is_fundamentally_a (type_instance=type_instance@entry=0x2451d90, fundamental_type=fundamental_type@entry=80) at ../gobject/gtype.c:4095
        node = <optimized out>
#1  0x00007f8fff720fae in g_object_ref (_object=0x2451d90) at ../gobject/gobject.c:3757
        _g_boolean_var_ = <optimized out>
        object = 0x2451d90
        old_val = <optimized out>
        object_already_finalized = <optimized out>
        __func__ = "g_object_ref"
#2  0x00007f8fff8e0bcb in _g_object_ref0 (self=<optimized out>) at libcore/libpantheon-files-core.so.6.3.0.p/Directory.c:1038
No locals.
#3  files_directory_cache_lookup (file=0x25082c0) at libcore/libpantheon-files-core.so.6.3.0.p/Directory.c:4313
        _tmp2_ = <optimized out>
        _tmp3_ = <optimized out>
        _tmp4_ = <optimized out>
        _tmp1_ = <optimized out>
        cached_dir = 0x0
        _tmp0_ = <optimized out>
        _tmp6_ = <optimized out>
        _inner_error0_ = 0x0
        result = <optimized out>
#4  0x00007f8fff8e4753 in _files_file___lambda8_ (self=0x7f8fec010800) at libcore/libpantheon-files-core.so.6.3.0.p/File.c:3952
        dir = 0x0
        _tmp1_ = <optimized out>
        _tmp3_ = 0
        _tmp2_ = <optimized out>
        _tmp4_ = <optimized out>
        _tmp0_ = <optimized out>
        _tmp0_ = <optimized out>
        dir = <optimized out>
        _tmp1_ = <optimized out>
        _tmp2_ = <optimized out>
        _tmp3_ = <optimized out>
        _tmp4_ = <optimized out>
        _tmp5_ = <optimized out>
        _tmp6_ = <optimized out>
        _tmp7_ = <optimized out>
        _tmp8_ = <optimized out>
        _tmp9_ = <optimized out>
        _tmp10_ = <optimized out>
#5  __files_file___lambda8__files_file_icon_changed (_sender=<optimized out>, self=0x7f8fec010800) at libcore/libpantheon-files-core.so.6.3.0.p/File.c:3988
No locals.
#6  0x00007f8fff71c420 in g_closure_invoke (closure=0x2d2fa30, return_value=return_value@entry=0x0, n_param_values=1, param_values=param_values@entry=0x7ffd88bf1050, 
    invocation_hint=invocation_hint@entry=0x7ffd88bf0fd0) at ../gobject/gclosure.c:832
        marshal = 0x7f8fff71e350 <g_cclosure_marshal_VOID__VOID>
        marshal_data = 0x0
        in_marshal = 0
        real_closure = 0x2d2fa10
        __func__ = "g_closure_invoke"
#7  0x00007f8fff72f3d6 in signal_emit_unlocked_R (node=node@entry=0x2d2f930, detail=detail@entry=0, instance=instance@entry=0x7f8fec010800, emission_return=emission_return@entry=0x0, 
    instance_and_params=instance_and_params@entry=0x7ffd88bf1050) at ../gobject/gsignal.c:3796
        tmp = <optimized out>
        handler = 0x2d7eb80
        accumulator = 0x0
        emission = {next = 0x0, instance = 0x7f8fec010800, ihint = {signal_id = 504, detail = 0, run_type = (G_SIGNAL_RUN_FIRST | G_SIGNAL_ACCUMULATOR_FIRST_RUN)}, state = EMISSION_RUN, 
          chain_type = 4}
        class_closure = 0x0
        hlist = <optimized out>
        handler_list = 0x2d7eb80
        return_accu = 0x0
        accu = {g_type = 0, data = {{v_int = 0, v_uint = 0, v_long = 0, v_ulong = 0, v_int64 = 0, v_uint64 = 0, v_float = 0, v_double = 0, v_pointer = 0x0}, {v_int = 0, v_uint = 0, 
              v_long = 0, v_ulong = 0, v_int64 = 0, v_uint64 = 0, v_float = 0, v_double = 0, v_pointer = 0x0}}}
        signal_id = 504
        max_sequential_handler_number = 4914
        return_value_altered = <optimized out>
        EMIT_RESTART = <optimized out>
#8  0x00007f8fff7362ed in g_signal_emit_valist (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>, var_args=var_args@entry=0x7ffd88bf11d0)
    at ../gobject/gsignal.c:3549
        instance_and_params = 0x7ffd88bf1050
        signal_return_type = <optimized out>
        param_values = 0x7ffd88bf1068
        node = <optimized out>
        i = <optimized out>
        n_params = <optimized out>
        __func__ = "g_signal_emit_valist"
#9  0x00007f8fff73656f in g_signal_emit (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>) at ../gobject/gsignal.c:3606
        var_args = {{gp_offset = 24, fp_offset = 48, overflow_arg_area = 0x7ffd88bf12b0, reg_save_area = 0x7ffd88bf11f0}}
#10 0x00007f8fff8e8078 in files_file_update (self=0x7f8fec010800) at libcore/libpantheon-files-core.so.6.3.0.p/File.c:1869
        _tmp0_ = <optimized out>
        _tmp1_ = <optimized out>
        _tmp2_ = <optimized out>
        _tmp4_ = <optimized out>
        _tmp5_ = <optimized out>
        _tmp6_ = <optimized out>
        _tmp14_ = <optimized out>
        target_uri = <optimized out>
        _tmp18_ = <optimized out>
        _tmp19_ = <optimized out>
        _tmp20_ = <optimized out>
        _tmp72_ = <optimized out>
        _tmp73_ = <optimized out>
        _tmp97_ = <optimized out>
        _tmp108_ = <optimized out>
        _tmp109_ = <optimized out>
        _tmp110_ = <optimized out>
        _tmp111_ = <optimized out>
        _tmp112_ = <optimized out>
        _tmp113_ = <optimized out>
        _tmp114_ = <optimized out>
        _tmp115_ = <optimized out>
        _tmp116_ = <optimized out>
        _tmp117_ = <optimized out>
        _tmp118_ = <optimized out>
        _tmp119_ = <optimized out>
        _tmp125_ = <optimized out>
        _tmp131_ = <optimized out>
        _inner_error0_ = 0x0
        __func__ = "files_file_update"
#11 0x00007f8fff8dc351 in files_directory_try_query_info_co (_data_=0x2f9cd20) at libcore/libpantheon-files-core.so.6.3.0.p/Directory.c:1537
        __func__ = "files_directory_try_query_info_co"
#12 0x00007f8fff5d4103 in g_task_return_now (task=task@entry=0x7f8fe004d8d0) at ../gio/gtask.c:1232
No locals.
#13 0x00007f8fff5d4ae3 in g_task_return (type=<optimized out>, task=0x7f8fe004d8d0) at ../gio/gtask.c:1301
        source = 0x7f8f880025d0
        source = <optimized out>
        source_name = <optimized out>
#14 g_task_return (task=0x7f8fe004d8d0, type=<optimized out>) at ../gio/gtask.c:1258
        source = <optimized out>
        source_name = <optimized out>
#15 0x00007f8fff8dc041 in files_directory_query_info_async_co (_data_=0x2443590) at libcore/libpantheon-files-core.so.6.3.0.p/Directory.c:3574
        __func__ = "files_directory_query_info_async_co"
#16 0x00007f8fff5d4103 in g_task_return_now (task=task@entry=0x2368c80) at ../gio/gtask.c:1232
No locals.
#17 0x00007f8fff5d4139 in complete_in_idle_cb (task=0x2368c80) at ../gio/gtask.c:1246
No locals.
#18 0x00007f8fff7be4df in g_main_dispatch (context=0x231e9f0) at ../glib/gmain.c:3454
        dispatch = 0x7f8fff7ba6a0 <g_idle_dispatch>
        prev_source = 0x0
        begin_time_nsec = 0
        was_in_call = 0
        user_data = 0x2368c80
        callback = 0x7f8fff5d4130 <complete_in_idle_cb>
        cb_funcs = <optimized out>
        cb_data = 0x3057e80
        need_destroy = <optimized out>
        source = 0x7f8f880025d0
        current = 0x235f100
        i = 2
        __func__ = "g_main_dispatch"
#19 g_main_context_dispatch (context=context@entry=0x231e9f0) at ../glib/gmain.c:4172
No locals.
#20 0x00007f8fff7be898 in g_main_context_iterate (context=context@entry=0x231e9f0, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/gmain.c:4248
        max_priority = 0
        timeout = 0
        some_ready = 1
        nfds = <optimized out>
        allocated_nfds = 3
        fds = 0x2d26df0
        begin_time_nsec = 0
#21 0x00007f8fff7be92c in g_main_context_iteration (context=context@entry=0x231e9f0, may_block=may_block@entry=1) at ../glib/gmain.c:4313
        retval = <optimized out>
#22 0x00007f8fff601f5d in g_application_run (application=application@entry=0x235a140, argc=argc@entry=1, argv=argv@entry=0x7ffd88bf1638) at ../gio/gapplication.c:2573
        arguments = 0x235c0a0
        status = 0
        context = 0x231e9f0
        acquired_context = <optimized out>
        __func__ = "g_application_run"
#23 0x000000000042fadc in _vala_main (args_length1=1, args=0x7ffd88bf1638) at src/io.elementary.files.p/main.c:72
        application = 0x235a140
        _tmp0_ = 0x235a140
        result = <optimized out>
        application = <optimized out>
        _tmp0_ = <optimized out>
        result = <optimized out>
#24 main (argc=1, argv=0x7ffd88bf1638) at src/io.elementary.files.p/main.c:81
No locals.
```

### Hardware Info

_No response_

### PR DESCRIPTION
**Pull request by jeremypw**
Fix #2148 (hopefully)

 - Directory cache now holds a toggle reference to the Directory 
 - Move Directory creation related methods together at top of file
 - Ensure  directories temporarily cached during creation are removed from cache if necessary
 - Amend test
 - Remove unnecessary null check (cache_lookup should never be called with a null parameter); Amend calling objects to comply
 - Add additional tests for directory_cache operations
 
A separate critical warning during Directory destruction is also fixed.

It was noticed that the `creation_key` property may no longer be required after https://github.com/elementary/files/commit/ef93c748789b0bcdd4b20565dff8e52a56993220.  The creation_key should be the same as the location now as it is always a folder and the location should no longer change during initialisation.

However, it was retained for now in case there is a corner case that needs it.

### CODE REVIEW

**Comment by bobby285271** [comment]
Yeah I can no longer reproduce the crash with this branch :partying_face:

File: libcore/Directory.vala [hunk 1]
```diff
@@ -531,14 +579,13 @@ public class Files.Directory : Object {
 
     private static void toggle_ref_notify (void* data, Object object, bool is_last) {
         if (is_last) {
-            unowned Directory dir = (Directory) object;
-            debug ("Directory is last toggle_ref_notify %s", dir.file.uri);
-
+            var dir = (Directory) object;
```
Comment (tintou) [inline libcore/Directory.vala:582]:
```suggestion
            unowned var dir = (Directory) object;
```

**Comment by jeremypw** [comment]
Just checking that all calls to cache_lookup can be guaranteed non-null.

File: libcore/Directory.vala (hunk 1 above)
Comment (jeremypw) [reply libcore/Directory.vala:582]: Can we keep the owned var and remove the dir.ref () instead? Is it equivalent?

**Comment by jeremypw** [comment]
I did manage to get a warning about the creation_key differing from the location when connected to a remote public WebDAV server but then I could not reproduce it.  The code should in any case be robust in this situation.

**Review by tintou — APPROVED** [review]

**Review by bobby285271 — APPROVED** [review]
Fixes the issue

**Pull request merged**